### PR TITLE
[Shell] Add FlyoutIcon 

### LINKED ISSUE #4766: [Shell] Cannot set hamburger button asset
### Description

Related to #3124 I can set the back button asset, but I cannot specify my own asset. The `3bar.png` is not readily obvious, and it would be much easier to specify my own asset in the `Shell.xaml` than have to figure that out.

### Basic Information

- Version with issue: 4.0-pre1
Parent: #2415

### LINKED ISSUE #4767: [Shell] BackButtonBehavior is used to override the icon
### Description

This is done to override the back button.  

```
var bbb = new BackButtonBehavior()
            {
                IconOverride = ImageSource.FromFile("close")
            };

            Shell.SetBackButtonBehavior(this, bbb);
```

"Behavior" here seems odd to me. Doesn't seem like a behavior. 

Also when I do this, it replaces the hamburger, but also doesn't open the menu.

### Expected Behavior

I can set the icon on a property of the Shell, a Style, or on the ContentPage.

### Actual Behavior

I can only set in code like above via a behavior.

### Basic Information

- Version with issue: 4.0-pre1

### Screenshots

https://www.screencast.com/t/TFrhknPNYhz3
Parent: #2415

### LINKED ISSUE #4845: [Shell] Back button behavior failures
### Description

Related to #4767 

My goal is to simply change the back button text, however it's more involved in Shell than when using NavigationPage where you can simply do this:

```
Xamarin.Forms.NavigationPage.SetBackButtonTitle(this, "");
```

**Problem 1: Cannot get the existing BackButtonBehavior**

This is always null:
```
var btn = Shell.GetBackButtonBehavior(this);
```

**Problem 2: To set the title and/or icon means all or nothing on the back button.  You have to also  implement the navigation through the command.**

```
Shell.SetBackButtonBehavior(this, new BackButtonBehavior
            {
                TextOverride = string.Empty,
                IconOverride = ImageSource.FromFile("icon_back_arrow"),
                
            });
```

The above results in a button visually that I want, but it doesn't do anything because the action is wrapped up in the behavior as well.

**Problem 3: When also setting the command, Shell crashes.**

```
Shell.SetBackButtonBehavior(this, new BackButtonBehavior
            {
                TextOverride = string.Empty,
                IconOverride = ImageSource.FromFile("icon_back_arrow"),
                Command = new Command(() => { Navigation.PopAsync(true); })
                
            });
```

The navigation pop works, but once back on the previous page it crashes.

```
Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at Xamarin.Forms.Platform.iOS.ShellPageRendererTracker+<UpdateToolbarItems>d__35.MoveNext () [0x000a2] in <d746bc31aa28499cbbacaa2963ef26ed>:0 
--- End of stack trace from previous location where exception was thrown ---
  at Xamarin.Forms.Platform.iOS.ShellPageRendererTracker+<SetBackButtonBehavior>d__38.MoveNext () [0x00116] in <d746bc31aa28499cbbacaa2963ef26ed>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_0 (System.Object state) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.2.1.12/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1023 
  at Foundation.NSAsyncSynchronizationContextDispatcher.Apply () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.2.1.12/src/Xamarin.iOS/Foundation/NSAction.cs:178 
--- End of stack trace from previous location where exception was thrown ---
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.2.1.12/src/Xamarin.iOS/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.2.1.12/src/Xamarin.iOS/UIKit/UIApplication.cs:63 
  at HowYouSay.iOS.Application.Main (System.String[] args) [0x00001] in /work/HowYouSay/iOS/Main.cs:17
2018-12-21 10:03:40.375558-0600 HowYouSay.iOS[24877:940095] Unhandled managed exception:
Object reference not set to an instance of an object (System.NullReferenceException)
  at Xamarin.Forms.Platform.iOS.ShellPageRendererTracker+<UpdateToolbarItems>d__35.MoveNext () [0x000a2] in <d746bc31aa28499cbbacaa2963ef26ed>:0 
--- End of stack trace from previous location where exception was thrown ---
  at Xamarin.Forms.Platform.iOS.ShellPageRendererTracker+<SetBackButtonBehavior>d__38.MoveNext () [0x00116] in <d746bc31aa28499cbbacaa2963ef26ed>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_0 (System.Object state) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.2.1.12/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1023 
  at Foundation.NSAsyncSynchronizationContextDispatcher.Apply () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.2.1.12/src/Xamarin.iOS/Foundation/NSAction.cs:178 
--- End of stack trace from previous location where exception was thrown ---
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.2.1.12/src/Xamarin.iOS/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.2.1.12/src/Xamarin.iOS/UIKit/UIApplication.cs:63 
  at HowYouSay.iOS.Application.Main (System.String[] args) [0x00001] in /work/HowYouSay/iOS/Main.cs:17
```

### Expected Behavior

I should be able to amend the title and/or icon without having to reimplement the navigation action itself.

### Actual Behavior

I have to do everything.

### Basic Information

- Version with issue: 4.0-pre1

Parent: #2415

### LINKED ISSUE #5219: [Shell,Android] BackButtonBehavior enables FlyoutBehavior
### Description

When the `FlyoutBehavior="Disabled"` and then I add a `BackButtonBehavior` on a `ContentPage`, the flyout becomes active on Android.

Android: https://www.screencast.com/t/xqjM4PeTrEH

iOS: https://www.screencast.com/t/VpYNt6aRJsj3

### Steps to Reproduce

```
<?xml version="1.0" encoding="UTF-8"?>
<Shell xmlns="http://xamarin.com/schemas/2014/forms" 
       xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml" 
       xmlns:local="clr-namespace:VisualChallenge"
       RouteHost="companyname.com"
       RouteScheme="app"
       Route="VisualChallenge"
       FlyoutBehavior="Disabled"
       Title="VisualChallenge"
       x:Class="VisualChallenge.AppShell"
       Visual="Material"
       >
```

```
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:sys="clr-namespace:System;assembly=mscorlib"
             x:Class="VisualChallenge.VisualChallengePage"
             Shell.NavBarIsVisible="true"
             Title="Edit Request"
             >
    
    <ContentPage.ToolbarItems>
        <ToolbarItem Text="Save"/>
    </ContentPage.ToolbarItems>
    
    <Shell.BackButtonBehavior>
        <BackButtonBehavior TextOverride="Undo" IconOverride="{OnPlatform Android=blanky}"></BackButtonBehavior>
    </Shell.BackButtonBehavior>
```

### Expected Behavior

Flyout doesn't become active.

### Basic Information

- Version with issue: 3.6-pre1
- Android 26
Parent: #2415

### PR DESCRIPTION
**Pull request by rmarinho**
### Description of Change ###

Fixes some issues related with BackButtonBehavior on Android and iOS. 
Provides new API for Shell.FlyoutIcon

### Issues Resolved ### 

- helps #3124 (Not yet fixed in this pr, provides a workaround)
- fixes #4766
- fixes #4767
- fixes #4845
- fixes #5219

### API Changes ###

Added:
public static readonly BindableProperty FlyoutIconProperty =
			BindableProperty.Create(nameof(FlyoutIcon), typeof(ImageSource), typeof(Shell), null);

### Platforms Affected ### 

- Core/XAML (all platforms)
- iOS
- Android

### Behavioral/Visual Changes ###

User can specify it's own icon 

### Before/After Screenshots ### 

Simulator Screen Shot - iPhone XS Max - 2019-03-15 at 12 09 00
Screenshot_1552651715

### PR Checklist ###

- [ ] Has automated tests 
- [ ] Rebased on top of the target branch at time of PR
- [ ] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Controls/XamStore/StoreShell.xaml.cs [hunk 1]
```diff
@@ -12,12 +12,13 @@ namespace Xamarin.Forms.Controls.XamStore
...
-		{
-			InitializeComponent ();
+			public StoreShell ()
+			{
+				InitializeComponent ();
 
-			CurrentItem = _storeItem;
-		}
+				CurrentItem =  _storeItem;
+				Routing.RegisterRoute("demo", typeof(DemoShellPage));
```
Comment (paymicro) [inline Xamarin.Forms.Controls/XamStore/StoreShell.xaml.cs]: It looks like it's from PR https://github.com/xamarin/Xamarin.Forms/pull/5564.

Comment (rmarinho) [reply Xamarin.Forms.Controls/XamStore/StoreShell.xaml.cs]: Yeah we need to review that one first please

File: Xamarin.Forms.Platform.Android/Renderers/ShellToolbarTracker.cs [hunk 2]
```diff
@@ -283,6 +286,27 @@ protected virtual async void UpdateLeftBarButtonItem(Context context, Toolbar to
...
+				if (item is IShellController shell)
+				{
+					icon = shell.FlyoutIcon;
+					if(icon != null)
+					{
+						var drawable = await context.GetFormsDrawable(icon);
+						actionBarDrawerToggle.DrawerArrowDrawable = new FlyoutIconDrawerDrawable(context, drawable);
+					}
+
+					item = null;
```
Comment (paymicro) [inline Xamarin.Forms.Platform.Android/Renderers/ShellToolbarTracker.cs]: why not use `break;` or `return;` here?

File: Xamarin.Forms.Platform.Android/Renderers/ShellToolbarTracker.cs [hunk 3]
```diff
@@ -283,6 +286,27 @@ protected virtual async void UpdateLeftBarButtonItem(Context context, Toolbar to
...
+			ImageSource icon = null;
+			while (!Application.IsApplicationOrNull(item))
+			{
+				if (item is IShellController shell)
+				{
+					icon = shell.FlyoutIcon;
+					if(icon != null)
+					{
+						var drawable = await context.GetFormsDrawable(icon);
+						actionBarDrawerToggle.DrawerArrowDrawable = new FlyoutIconDrawerDrawable(context, drawable);
```
Comment (paymicro) [inline Xamarin.Forms.Platform.Android/Renderers/ShellToolbarTracker.cs]: I think here need add a null check for `drawable` or add support null drawable in `FlyoutIconDrawerDrawable` class

**Comment by samhouts** [comment]
@rmarinho Does this fix #5646?

**Comment by samhouts** [comment]
@rmarinho Does this fix #3124?

File: Xamarin.Forms.Platform.iOS/Renderers/ShellPageRendererTracker.cs [hunk 4]
```diff
@@ -210,16 +213,34 @@ protected virtual async Task UpdateToolbarItems()
...
+			{
+				if (item is IShellController shell)
+				{
+					image = shell.FlyoutIcon;
+					item = null;
+				}
+				item = item?.Parent;
+			}
+			if (image == null)
+				image = "3bar.png";
```
Comment (paymicro) [inline Xamarin.Forms.Platform.iOS/Renderers/ShellPageRendererTracker.cs]: Perhaps `image = ImageSource.FromResource("3bar.png");`

File: Xamarin.Forms.Platform.Android/Renderers/ShellToolbarTracker.cs [hunk 5]
```diff
@@ -235,51 +236,114 @@ protected virtual async void UpdateLeftBarButtonItem(Context context, Toolbar to
...
+			var text = behavior.TextOverride;
+			var enabled = behavior.IsEnabled;
+
+			Drawable icon = null;
+
+			if (image != null)
+				icon = await context.GetFormsDrawable(image);
+
+			if (text != null)
+				icon = new FlyoutIconDrawerDrawable(context, TintColor, null, text);
```
Comment (paymicro) [inline Xamarin.Forms.Platform.Android/Renderers/ShellToolbarTracker.cs]: `TextOverride` has higher priority over `IconOverride`. I think they can be used at the same time.
```suggestion
				icon = new FlyoutIconDrawerDrawable(context, TintColor, icon, text);
```

File: Xamarin.Forms.Controls/XamStore/Views/StorePages.cs [hunk 6]
```diff
@@ -209,13 +209,29 @@ public BasePage(string title, Color tint)
...
 
+			grid.Children.Add(MakeButton("Navigate to 'demo' route",
+				async () => await Shell.CurrentShell.GoToAsync("demo", true)),
+			1, 15);
+
+			grid.Children.Add(MakeButton("Go Back with Text",
+			async () => {
+					var page = (Page)Activator.CreateInstance(GetType());
+					Shell.SetShellForegroundColor(page, Color.Pink);
+					Shell.SetBackButtonBehavior(page, new BackButtonBehavior()
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply Xamarin.Forms.Controls/XamStore/Views/StorePages.cs]: good catch

File: Xamarin.Forms.Platform.Android/Renderers/ShellToolbarTracker.cs [hunk 7]
```diff
@@ -452,5 +516,51 @@ void UpdateToolbarItems()
...
+				if (_iconBitmap != null)
+				{
+					_iconBitmap.SetBounds(Bounds.Left, Bounds.Top, Bounds.Right, Bounds.Bottom);
+					_iconBitmap.Draw(canvas);
+				}
+				else if (!string.IsNullOrEmpty(_text))
+				{
+					var paint = new Paint { AntiAlias = true };
+					paint.TextSize = _defaultSize;
+					paint.Color = pressed ? _pressedBackgroundColor.ToAndroid() : _defaultColor.ToAndroid();
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply Xamarin.Forms.Platform.Android/Renderers/ShellToolbarTracker.cs]: right, @paymicro  @PureWeen @StephaneDelcroix  anyone has a good idea for this ? should we go all in and implement states or use a button?  I might be overthink this.

File: Xamarin.Forms.Platform.iOS/Renderers/ShellPageRendererTracker.cs (hunk 4 above)
Comment (rmarinho) [reply Xamarin.Forms.Platform.iOS/Renderers/ShellPageRendererTracker.cs]: we need something better, leaving for now, but we might want to draw the 3 bars

Comment (rmarinho) [reply Xamarin.Forms.Platform.iOS/Renderers/ShellPageRendererTracker.cs]: let's track this when e try fix for good #3124

**Pull request merged**